File: src/s0003_longest_substring_without_repeating_characters.rs
/**
 * [3] Longest Substring Without Repeating Characters
 *
 * Given a string, find the length of the longest substring without repeating characters.
 *
 * Example:
 *
 * Input: "abcabcbb"
 * Output: 3
 * Explanation: The answer is "abc", with the length of 3.
 *
 */
use std::collections::HashSet;

pub struct Solution {}

impl Solution {
    pub fn length_of_longest_substring(s: &str) -> u32 {
        let s_vec: Vec<char> = s.chars().into_iter().collect();
        let mut lookup = HashSet::new();
        let (mut left, mut max_len, mut right) = (0, 0, 0);
        while right < s_vec.len() {
            let c = s_vec[right];
            right += 1;

            println!("[+] window: [{}, {})", left, right);
            while !lookup.insert(c) {
                // flag = true;
                let d = s_vec[left];
                lookup.remove(&d);
                left += 1;
            }
            max_len = std::cmp::max(max_len, right - left);     // update max len here guarantee all ele can insert into lookup

        }
        max_len as u32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    // chars(&self)
    fn test_string_methods() {
        let word = "goodbye";

        let count = word.chars().count();
        assert_eq!(7, count);

        let mut word = "hello".chars();
        assert_eq!(Some('h'), word.next());
        assert_eq!(Some('e'), word.next());
        assert_eq!(Some('l'), word.next());
        assert_eq!(Some('l'), word.next());
        assert_eq!(Some('o'), word.next());
    }

    #[test]
    // for_each<F>(self, f: F)
    // where
    // F: FnMut(Self::Item), 
    fn test_iterator_trait() {
        let v: Vec<_> =vec![0, 1, 2, 3, 4, 5];
        v.iter().enumerate().for_each(|(i, v)| {
            assert_eq!(0, *v);
        });
    }

    #[test]
    fn test_3() {
        assert_eq!(3, Solution::length_of_longest_substring("abcabcbb"));
        assert_eq!(2, Solution::length_of_longest_substring("ab"));
        assert_eq!(Solution::length_of_longest_substring("bbbb"), 1);
        assert_eq!(
            Solution::length_of_longest_substring("pwwkew"),
            3
        );
    }
}

File: cpp/sliding-window/s0003_longest_substring_without_repeating_characters.cpp
// problem: https://leetcode.cn/problems/longest-substring-without-repeating-characters
// 示例 1:
// 输入: s = "abcabcbb"
// 输出: 3 
// 解释: 因为无重复字符的最长子串是 "abc", 所以其长度为 3.
// 
// 示例 2:
// 输入: s = "bbbbb"
// 输出: 1
// 解释: 因为无重复字符的最长子串是 "b", 所以其长度为 1.
// 
// 示例 3:
// 输入: s = "pwwkew"
// 输出: 3
// 解释: 因为无重复字符的最长子串是 "wke", 所以其长度为 3.
//      请注意, 你的答案必须是子串的长度, "pwke" 是一个子序列, 不是子串.
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.empty()) return 0;
        int l = s.size();
        int left = 0, right = 0;
        unordered_set<char> window;
        int max_len = INT_MIN;

        while(right < l) {
            char rc = s[right];
            right ++;
            while (window.count(rc))
            {
                char lc = s[left];
                left ++;
                window.erase(lc);
            }
            window.emplace(rc);
            max_len = max(max_len, right - left);
        }

        return max_len;
    }
};

int main() 
{
    Solution s;
    int ret = s.lengthOfLongestSubstring("abcabcbb");
    assert(ret == 3);

    return 0;
}